R/modify-records.R

```diff
@@ -323,6 +323,131 @@ get_table_columns <- function(.mod, from_data = TRUE){
+  expr <- gsub("\\.LE\\.", "<=", expr)
+  expr <- gsub("\\.GT\\.", ">", expr)
+  expr <- gsub("\\.GE\\.", ">=", expr)
+  return(expr)
+}
+
+# Function to invert an expression (for ignore statements)
+invert_expression <- function(expr) {
+  expr <- dplyr::case_when(
+    grepl("==", expr) ~ gsub("==", "!=", expr),
+    grepl("!=", expr) ~ gsub("!=", "==", expr),
+    grepl("<=", expr) ~ gsub("<=", ">=", expr),
+    grepl(">=", expr) ~ gsub(">=", "<=", expr),
+    grepl("<", expr) ~ gsub("<", ">", expr),
+    grepl(">", expr) ~ gsub(">", "<", expr),
+    TRUE ~ expr
+  )
+  return(expr)
+}
+
+
+#' Filter `NONMEM` input data based on `IGNORE` and `ACCEPT` record options
+#'
+#' @param .mod a `bbi_nonmem_model` object
+#' @param data a starting dataset
+#' @keywords internal
+filter_nm_data <- function(.mod, data = nm_data(.mod)){
+
+  data_recs <- get_records(.mod, "data")
+  n_data <- length(data_recs)
+  if(n_data !=1){
+    recs_fmt <- purrr::map_chr(data_recs, function(rec) rec$format())
+    rlang::abort(
+      c(
+        glue::glue("Expected a single data record, but found {n_data}:\n\n"),
+        recs_fmt
+      )
+    )
+  }
+  data_recs[[1]]$parse()
+
+  # Extract & format IGNORE options
+  ignore_opts <- purrr::keep(data_recs[[1]]$values, function(val){
+    inherits(val, "nmrec_option_value") && identical(val[["name"]], "ignore")
+  })
+  ignore_vals <- purrr::map_chr(ignore_opts, function(ign_val){
+    gsub("\\(|\\)", "", unquote_filename(ign_val$value))
+  })
+
+  # Extract & format ACCEPT options
+  accept_opts <- purrr::keep(data_recs[[1]]$values, function(val){
+    inherits(val, "nmrec_option_value") && identical(val[["name"]], "accept")
+  })
+  accept_vals <- purrr::map_chr(accept_opts, function(acc_val){
+    gsub("\\(|\\)", "", unquote_filename(acc_val$value))
+  })
+
+
+  # Translate the ignore and accept expressions
+  ignore_exprs <- lapply(ignore_vals, translate_operator) %>% gsub(",", " &", .)
+  accept_exprs <- lapply(accept_vals, translate_operator) %>% gsub(",", " &", .)
+
+  ## Combine the expressions into filter expressions ##
+  data_cols <- get_input_columns(.mod)
+
+  # `IGNORE=#`, `IGNORE=@`, `IGNORE=c1`, `IGNORE=(list)`
+  ignore_filters <- purrr::map_chr(ignore_exprs, function(expr) {
+    if(expr == "#"){
+      # IGNORE=# is the default.  That is, in the absence of IGNORE option, any
+      # record whose first character is '#' is treated as a comment record.
+      col_filters <- purrr::map_chr(data_cols, function(col) {
+        paste0("!grepl('^#', ", col, ")")
+      })
+      return(paste(col_filters, collapse = " & "))
+    }else if(expr == "@"){
+      # IGNORE=@ signifies that any data record having an alphabetic character
+      # or `@` as its first non-blank character (not just in column 1)
+      # should be ignored. This permits a table file having header lines to be
+      # used as an NM-TRAN data set.
+      col_filters <- purrr::map_chr(data_cols, function(col) {
```

Comment:
The documentation says "not just in column 1", but for this application I think we are still looking at column 1; I don't know how looking at all data cols will work here.

Reply:
The filters for `inst/models/nonmem/basic/1.ctl` come back as: 

```
[1] "!grepl('^[A-Za-z@]', ID) & !grepl('^[A-Za-z@]', TIME) & !grepl('^[A-Za-z@]', MDV) & !grepl('^[A-Za-z@]', EVID) & !grepl('^[A-Za-z@]', DV) & !grepl('^[A-Za-z@]', AMT) & !grepl('^[A-Za-z@]', SEX) & !grepl('^[A-Za-z@]', WT) & !grepl('^[A-Za-z@]', ETN) & !grepl('^[A-Za-z@]', NUM)"
[2] "ID!=2"     
```

The first filter will drop any row that contains a column with alpha or `@`.

Reply:
I think this behavior is intended if you apply the `@` filter though right? As you affirm, the `NONMEM` documentation does say the `@` applies to all columns

Reply:
@kylebaron I actually just discovered something that appears to be a `NONMEM` bug. I processed the `1.ctl` example using the defaults, and compared it to `FDATA` and was surprised that the first record was filtered out _in addition_ to the header row:

```r
# FDATA - captured using the NM-TRAN PR (not merged yet)
> nmtran_p <- run_nmtran(.mod, delete_on_exit = F)
> fdata <- nm_file_impl(file.path(nmtran_p$run_dir, "FDATA")) %>% stats::setNames(names(nm_data(.mod)))

# R-processed NM-TRAN data
> nmtran_data <- nm_data_as_nmtran(.mod, data = data)


> fdata
# A tibble: 778 × 10
      ID  TIME   MDV  EVID    DV   AMT   SEX    WT   ETN   NUM
   <int> <dbl> <int> <int> <dbl> <int> <int> <dbl> <int> <int>
 1     1  0.25     0     0  12.6     0     1  51.6     1     2
 2     1  0.5      0     0  11.2     0     1  51.6     1     3
 3     1  0.75     0     0  17.7     0     1  51.6     1     4
 4     1  1        0     0  21.9     0     1  51.6     1     5
 5     1  1.25     0     0  16.2     0     1  51.6     1     6
 6     1  1.5      0     0  18.7     0     1  51.6     1     7
 7     1  1.75     0     0  23.8     0     1  51.6     1     8
 8     1  2        0     0  16.5     0     1  51.6     1     9
 9     1  2.25     0     0  17.3     0     1  51.6     1    10
10     1  2.5      0     0  18.4     0     1  51.6     1    11
# ℹ 768 more rows
# ℹ Use `print(n = ...)` to see more rows

> nmtran_data
# A tibble: 779 × 10
      ID  TIME   MDV  EVID    DV   AMT   SEX    WT   ETN   NUM
   <int> <dbl> <int> <int> <dbl> <int> <int> <dbl> <int> <int>
 1     1  0        0     0  1.22     0     1  51.6     1     1
 2     1  0.25     0     0 12.6      0     1  51.6     1     2
 3     1  0.5      0     0 11.2      0     1  51.6     1     3
 4     1  0.75     0     0 17.7      0     1  51.6     1     4
 5     1  1        0     0 21.9      0     1  51.6     1     5
 6     1  1.25     0     0 16.2      0     1  51.6     1     6
 7     1  1.5      0     0 18.7      0     1  51.6     1     7
 8     1  1.75     0     0 23.8      0     1  51.6     1     8
 9     1  2        0     0 16.5      0     1  51.6     1     9
10     1  2.25     0     0 17.3      0     1  51.6     1    10
# ℹ 769 more rows
# ℹ Use `print(n = ...)` to see more rows
```
You'll notice that the `TIME=0` record for `ID==1` was filtered out in `fdata`. This is _only_ true for the first ID too. This feels like a `NONMEM` bug, as I cant imagine filtering out that first record is an expectation.

Reply:
@barrettk  - I think records and columns refer to rows and columns _in the data file_, not in the realized data frame.  As currently implemented, finding `IGNORE=@` will drop all data rows where any data frame column contains alpha. I posted an example demonstrating this.

Reply:
> The filters for `inst/models/nonmem/basic/1.ctl` come back as:
> 
> ```
> [1] "!grepl('^[A-Za-z@]', ID) & !grepl('^[A-Za-z@]', TIME) & !grepl('^[A-Za-z@]', MDV) & !grepl('^[A-Za-z@]', EVID) & !grepl('^[A-Za-z@]', DV) & !grepl('^[A-Za-z@]', AMT) & !grepl('^[A-Za-z@]', SEX) & !grepl('^[A-Za-z@]', WT) & !grepl('^[A-Za-z@]', ETN) & !grepl('^[A-Za-z@]', NUM)"
> [2] "ID!=2"     
> ```
> 
> The first filter will drop any row that contains a column with alpha or `@`.
> 
> If you run this on `basic/1`, changing SEX from 0,1 to M,F, it will return no rows.

If you look at that set of filter statements, if you have _any_ data frame column that is character (`[A-Za-z]`) you won't get any of the rows back. This is not what is intended by `IGNORE=@`. The intended behavior is dropping any rows in the file that start with `[A-Za-z]`.

Reply:
I get NMTRAN errors when I try to reproduce your example though (truncated). The resulting FDATA is 19 records, all for `ID = 1`
```r
 WARNINGS AND ERRORS (IF ANY) FOR PROBLEM    1
             
 (WARNING  2) NM-TRAN INFERS THAT THE DATA ARE POPULATION.

 (DATA ERROR)      RECORD         1, DATA ITEM   7, CONTENTS: F
 ITEM IS NOT A NUMBER.

 (DATA ERROR)      RECORD         2, DATA ITEM   7, CONTENTS: F
 ITEM IS NOT A NUMBER.

 (DATA ERROR)      RECORD         3, DATA ITEM   7, CONTENTS: F
```

```r
> fdata
# A tibble: 18 × 10
      ID  TIME   MDV  EVID    DV   AMT SEX      WT   ETN   NUM
   <int> <dbl> <int> <int> <dbl> <int> <chr> <dbl> <int> <int>
 1     1  0.25     0     0 12.6      0 F      51.6     1     2
 2     1  0.5      0     0 11.2      0 F      51.6     1     3
 3     1  0.75     0     0 17.7      0 F      51.6     1     4
 4     1  1        0     0 21.9      0 F      51.6     1     5
 5     1  1.25     0     0 16.2      0 F      51.6     1     6
 6     1  1.5      0     0 18.7      0 F      51.6     1     7
 7     1  1.75     0     0 23.8      0 F      51.6     1     8
 8     1  2        0     0 16.5      0 F      51.6     1     9
 9     1  2.25     0     0 17.3      0 F      51.6     1    10
10     1  2.5      0     0 18.4      0 F      51.6     1    11
11     1  2.75     0     0 18.3      0 F      51.6     1    12
12     1  3        0     0 18.4      0 F      51.6     1    13
13     1  4        0     0 13.1      0 F      51.6     1    14
14     1  7        0     0 15.1      0 F      51.6     1    15
15     1 10        0     0  8.45     0 F      51.6     1    16
16     1 13        0     0 11.0      0 F      51.6     1    17
17     1 16        0     0  5.74     0 F      51.6     1    18
18     1 19        0     0  8.42     0 F      51.6     1    19
```

Input data for reference
```r
> data
# A tibble: 799 × 10
      ID  TIME   MDV  EVID    DV   AMT SEX      WT   ETN   NUM
   <int> <dbl> <int> <int> <dbl> <int> <chr> <dbl> <int> <int>
 1     1  0        0     0  1.22     0 F      51.6     1     1
 2     1  0.25     0     0 12.6      0 F      51.6     1     2
 3     1  0.5      0     0 11.2      0 F      51.6     1     3
 4     1  0.75     0     0 17.7      0 F      51.6     1     4
 5     1  1        0     0 21.9      0 F      51.6     1     5
 6     1  1.25     0     0 16.2      0 F      51.6     1     6
 7     1  1.5      0     0 18.7      0 F      51.6     1     7
 8     1  1.75     0     0 23.8      0 F      51.6     1     8
 9     1  2        0     0 16.5      0 F      51.6     1     9
10     1  2.25     0     0 17.3      0 F      51.6     1    10
# ℹ 789 more rows
```

Reply:
@barrettk  - please revisit `inst/model/nonmem/basic/2`; this has been updated to actually run this model (was just filtering last time).

Reply:
Again, you have to check this on `basic/2`; you won't see anything on `basic/1` because there is no non-numeric data in _the csv file_. 


```
4c4
< $DATA ../../../../extdata/acop.csv IGNORE=@
---
> $DATA ../../../../extdata/acop2.csv IGNORE=@
42,47c42,47
< $TABLE NUM IPRED NPDE CWRES NOPRINT ONEHEADER FILE=1.tab
< $TABLE NUM CL V KA ETAS(1:LAST) NOAPPEND NOPRINT ONEHEADER FILE=1par.tab
< $TABLE ID CL NOAPPEND NOPRINT ONEHEADER FIRSTONLY FILE=1first1.tab ; for testing firstonly
< $TABLE NUM CL NOAPPEND NOPRINT ONEHEADER FIRSTONLY FILE=1first2.tab ; for testing firstonly
< $TABLE NUM ID CL NOAPPEND NOPRINT ONEHEADER FIRSTONLY FILE=1first3.tab ; for testing firstonly
< $TABLE NUM CL KA FAKE NOAPPEND NOPRINT ONEHEADER FILE=1dups.tab ; for testing dropping duplicate columns
---
> $TABLE NUM IPRED NPDE CWRES NOPRINT ONEHEADER FILE=2.tab
> $TABLE NUM CL V KA ETAS(1:LAST) NOAPPEND NOPRINT ONEHEADER FILE=2par.tab
> $TABLE ID CL NOAPPEND NOPRINT ONEHEADER FIRSTONLY FILE=2first1.tab ; for testing firstonly
> $TABLE NUM CL NOAPPEND NOPRINT ONEHEADER FIRSTONLY FILE=2first2.tab ; for testing firstonly
> $TABLE NUM ID CL NOAPPEND NOPRINT ONEHEADER FIRSTONLY FILE=2first3.tab ; for testing firstonly
> $TABLE NUM CL KA FAKE NOAPPEND NOPRINT ONEHEADER FILE=2dups.tab ; for testing dropping duplicate columns
```